### PR TITLE
Add support for `body_path` in HTTP probe config.

### PR DESCRIPTION
**Pull request by emmanuel**
Fixes prometheus/blackbox_exporter#391.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -135,6 +139,13 @@ func (s *HTTPProbe) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if err := s.HTTPClientConfig.Validate(); err != nil {
 		return err
 	}
+	if s.BodyPath != "" {
```
Comment (brian-brazil) [inline config/config.go]: This should be opened and read in the request. It should also be an error for both to be specified.

File: config/config.go [hunk 2]
```diff
@@ -52,18 +54,20 @@ type Module struct {
...
+	IPProtocolFallback     bool              `yaml:"ip_protocol_fallback,omitempty"`
+	NoFollowRedirects      bool              `yaml:"no_follow_redirects,omitempty"`
+	FailIfSSL              bool              `yaml:"fail_if_ssl,omitempty"`
+	FailIfNotSSL           bool              `yaml:"fail_if_not_ssl,omitempty"`
+	Method                 string            `yaml:"method,omitempty"`
+	Headers                map[string]string `yaml:"headers,omitempty"`
+	FailIfMatchesRegexp    []string          `yaml:"fail_if_matches_regexp,omitempty"`
+	FailIfNotMatchesRegexp []string          `yaml:"fail_if_not_matches_regexp,omitempty"`
+	Body                   string            `yaml:"body,omitempty"`
+	BodyPath               string            `yaml:"body_path,omitempty"`
```
Comment (brian-brazil) [inline config/config.go]: body_file is the usual naming

File: config/config.go (hunk 1 above)
Comment (emmanuel) [reply config/config.go]: Updated to be opened within `ProbeHTTP`. I left the current `os.Open` call in `config.go` in order to fail fast on bad configs. This is different from `BearerToken`/`BearerTokenFile` (https://github.com/prometheus/common/blob/master/config/http_config.go#L81).

Let me know if I should drop opening the file during config validation.

File: config/config.go (hunk 2 above)
Comment (emmanuel) [reply config/config.go]: Renamed to `body_file`.

**Comment by emmanuel** [comment]
Added DCO. I'm not sure why the Travis build is failing, but from my (short) look, it appears related to git remotes, and not the content of these changes.

File: config/config.go [hunk 3]
```diff
@@ -135,6 +138,15 @@ func (s *HTTPProbe) UnmarshalYAML(unmarshal func(interface{}) error) error {
 	if err := s.HTTPClientConfig.Validate(); err != nil {
 		return err
 	}
+	if len(s.Body) > 0 && len(s.BodyFile) > 0 {
+		return errors.New("at most one of body & body_file may be configured")
+	}
+	if s.BodyFile != "" {
+		_, err := os.Open(s.BodyFile)
```
Comment (brian-brazil) [inline config/config.go]: You should close the file

File: prober/http.go [hunk 4]
```diff
@@ -257,10 +258,19 @@ func ProbeHTTP(ctx context.Context, target string, module config.Module, registr
...
 	var body io.Reader
 
-	// If a body is configured, add it to the request.
+	// If a body string is configured, add it to the request.
 	if httpConfig.Body != "" {
 		body = strings.NewReader(httpConfig.Body)
 	}
+	// If a body path is configured, add it to the request.
+	if httpConfig.BodyFile != "" {
+		file, err := os.Open(httpConfig.BodyFile)
```
Comment (brian-brazil) [inline prober/http.go]: Make sure this gets closed

File: prober/http_test.go [hunk 5]
```diff
@@ -164,6 +164,33 @@ func TestPost(t *testing.T) {
...
+	registry := prometheus.NewRegistry()
+	testCTX, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+	result := ProbeHTTP(testCTX, ts.URL,
+		config.Module{
+			Timeout: time.Second,
+			HTTP: config.HTTPProbe{
+				IPProtocolFallback: true,
+				Method:             "POST",
+				BodyFile:           "./http.go",
```
Comment (brian-brazil) [inline prober/http_test.go]: You can create a testdata directory, and a file with known contents there - plus verify that's what's coming through.

Comment (brian-brazil) [reply prober/http_test.go]: You should still change this to use explicit testdata files, rather than the source itself

**Comment by brian-brazil** [comment]
This has gone stale, let us know if you still wish to work on this.

**Pull request closed without merging**